Guidelines pre-PEP: Subset of C, some Types, Return values, Output arguments (#53)
capi-workgroup/api-evolution: guidelines.rst

    - https://github.com/capi-workgroup/api-evolution/issues/22

Portable subset of C
--------------------

While the C API is defined in terms of the C language, it supports building
wrappers for languages other than C, such as Rust, Java, assembly, or
Python with ``ctypes``.
These wrappers cannot realisically use a full-featured C parser.
To make the public API easier to describe and wrap, it should
avoid some of C's features:

*  Avoid ``static inline`` functions and macros.
   All functions must be exported as actual library symbols.
*  Avoid variadic functions.
*  Avoid C-specific types like ``long long``, ``enum`` or bit fields
   (see :ref:`types`).

Once you add API that conforms to this “portable subset”,
you can add additional C/C++-specific API. Usually, the additional API
will be either more performant, or easier to use from C.
For example:

*  A function may be shadowed by a ``static inline`` function or macro with
   the same behavior. Typically, this allows better performance for C/C++.
   See `shadowing example`_.
*  A function that uses C-specific types, such as ``PyLong_AsLongLong``,

>    is OK if equivalent functions are provided for the preferred types.

Reviewer: ```suggestion
   is OK if equivalent functions are provided for :ref:`the preferred types <types>`.
```

*  A variadic function is OK if there's an equivalent function

>    that takes an array.

Reviewer 2: ```
PyObject* PyObject_Call(PyObject *callable, PyObject *args, PyObject *kwargs)
PyObject* PyObject_CallFunction(PyObject *callable, const char *format, ...)
```
I do not see array here.

The closest analogues are
```
PyObject* PyObject_Vectorcall(PyObject *callable, PyObject *const *args, size_t nargsf, PyObject *kwnames)
PyObject* PyObject_VectorcallDict(PyObject *callable, PyObject *const *args, size_t nargsf, PyObject *kwdict)
```
And this (with addition of `PyObject_VectorcallMethod`) are only examples of such functions.

>    that takes an array.

Reviewer 3: `PyObject *args` is a tuple of arguments, which is an array-like alternative to the varargs in `CallFunction`.

We don't mean they must literally be a raw pointer to a sequential list of values. But there should be a reasonably efficient way for non-C code to call the function.

>    that takes an array.

Reviewer 2: I think that this is too wide interpretation of term "array", especially in the context of the C API. Could we use something less misleading?

>    that takes an array.

Reviewer 3: I'd be happy with "...an equivalent function that doesn't use varargs but allows the caller to pass all values in a single call, for example with an array". That might be a bit wordy though.

>    that takes an array.

Reviewer 4: When PySys_Audit() was added to the limited C API 3.13, `int PySys_AuditTuple(const char *event, PyObject *args)` was also added: pass an array of Python objects as a tuple.

> *  A variadic function is OK if there's an equivalent function
>    that takes an array.

Author: Keep in mind the WG may grant exceptions (esp. for consistency with existing API).

This is already a list of examples, but let's repeat that?
```suggestion
*  A variadic function is OK if there's an equivalent function
   that, for example, takes an array or tuple.
```

FWIW, if `PyUnicode_FromFormat` was added today, I *would* argue for a version that takes an array (of tagged unions).

>    that takes an array.

Reviewer 2: `PySys_Audit()` and `PySys_AuditTuple()` is a good example. `PySys_Audit() = Py_BuildValue() + PySys_AuditTuple()`.

Other examples are:
`PyErr_Format() = PyUnicode_FromFormat() + PyErr_SetObject()`
`PyErr_WarnFormat() = PyUnicode_FromFormat() + ... + PyErr_WarnExplicit()`
`PyErr_WarnExplicitFormat() = PyUnicode_FromFormat() + PyErr_WarnExplicit()`
`PyUnicodeWriter_Format() = PyUnicode_FromFormat() + PyUnicodeWriter_WriteStr()`

There equivalent functions take a string. So I would add a string to example.

Array cannot be used for heterogeneous arguments. Functions like `PyUnicode_FromFormat()` are C-specific. Other programming languages can implement equivalent function using native means or do not implement them at all.

> *  A variadic function is OK if there's an equivalent function
>    that takes an array.

Author: What about:
```suggestion
*  A variadic function is OK if there's an non-variadic function
   with equivalent functionality.
   Usually, the equivalent take an array or pre-constructed
   Python object: for example, `PyObject_Vectorcall` is the equivalent
   for `PyObject_CallFunction`; `PySys_AuditTuple` for `PySys_Audit`;
   `PyErr_SetObject` for `PyErr_Format`.
```

> Other programming languages can implement equivalent function using native means or do not implement them at all.

Yes. What's important that it's *possible* (and better, *easy*) to implement them by other means. 
For string formatting, `PyUnicodeWriter` helps :)

>    that takes an array.

Author: I agree, but the example is already long. It doesn't need to be complete :)

(Most of the `PyObject_Call*` functions are equivalent to one another -- i.e. redundant -- but that's a [different issue](https://github.com/capi-workgroup/api-evolution/issues/48).)

Macros can be used in the following cases:

*  Feature flags (e.g. ``HAVE_FORK``, ``Py_LIMITED_API``)
*  Simple constants (e.g. ``Py_TPFLAGS_BASETYPE``, ``PY_VERSION_HEX``).
*  Shortcuts for functionality that can be accomplished trivially,
   but perhaps tediously, without macros (e.g. ``Py_VISIT``,
   ``Py_BEGIN_ALLOW_THREADS``, ``Py_RETURN_RICHCOMPARE``).
   In this case, the macro-less equivalent should be clear from documentation:
   consider adding the macro's expansion to the docs.
   Non-C wrappers are expected re-implement these macros.
*  Features that aren't needed in non-C languages (e.g. ``Py_MAX``,
   ``Py_STRINGIFY``).
*  Macros used to define the API (e.g. ``PyAPI_FUNC``, ``Py_ALWAYS_INLINE``,
   ``Py_OBJECT_H``).
*  As an implementation detail (for example, when shadowing a function).

As always, new exceptions can be added with approval from the C API
working group.

.. note::

    For background and discussions, see:

>     - https://github.com/capi-workgroup/api-evolution/issues/11 (Treat the ABI as an API)
>     - https://github.com/capi-workgroup/api-evolution/issues/18 (Avoid macros and static inline functions)
>     - https://github.com/capi-workgroup/api-evolution/issues/12 (variadics)

Author: This is ReST, not Markdown.
I'll leave this kind of formatting for later, when it's time to turn this into a PEP.
But, “background and discussions” is a great place for the PEP 670 link:
```suggestion
    - :pep:`670`
    - https://github.com/capi-workgroup/api-evolution/issues/11 (Treat the ABI as an API)
    - https://github.com/capi-workgroup/api-evolution/issues/18 (Avoid macros and static inline functions)
    - https://github.com/capi-workgroup/api-evolution/issues/12 (variadics)
```

.. _types:

Types
=====

Arithmetic types
----------------

Avoid types with compiler-/platform-specific sizes, such as  ``long``
or ``unsigned short``.

Instead, use:

> *  ``int32_t`` and other C99+ fixed width integer types
> *  ``Py_ssize_t``, ``intptr_t``, ``ptrdiff_t`` for values of the appropriate
>    platform-specific types
> *  ``char``
> *  ``double`` (IEEE 754 ``binary64``)
> 
> The only exception is ``int``, which should be used for small ranges
> (typically, as a replacement for enum).
> If the 16-bit limit is relevant, and for all unsigned values, prefer
> explicit fixed width types over ``int``.

Author: > char can be signed or not, I would suggest to not use it in new APIs for arithmetic types.

Right. But, it's useful for strings.

```suggestion
*  ``int32_t`` and other C99+ fixed width integer types
*  ``Py_ssize_t``, ``intptr_t``, ``ptrdiff_t`` for values of the appropriate
   platform-specific types
*  ``double`` (IEEE 754 ``binary64``)

As an exception, use ``int`` for small ranges (typically, as a replacement
for enum).
If the 16-bit limit is relevant, and for all unsigned values, prefer
explicit fixed width types over ``int``.

As another exception, use ``char*`` for UTF-8/ASCII strings (and ``char``
for bytes of such text). For byte-strings, prefer ``uint8_t*`` or ``void*``.
```

For memory sizes and byte counts, use the signed ``Py_ssize_t``,
not the unsigned ``size_t``.

Enums and bitfields
-------------------

Avoid ``enum``, which have compiler-/platform-dependent size.
(CPython can not yet use C23's fixed underlying `enum` types.)
Instead, use ``int`` with defined constants.

> Also avoid bitfields, which have compiler-/platform-dependent memory layout.

Reviewer: ```suggestion
Avoid bitfields, which have compiler-/platform-dependent memory layout.
```

Instead, use fixed width integer types and bitmask constants.

> To be clear: it is fine to use ``enum`` and bitfields outside the
> public headers.

Reviewer 3: ```suggestion
To be clear: it is fine to use ``enum`` and bitfields within CPython sources,
provided they do not need to be used in public headers.
```

    - https://github.com/capi-workgroup/api-evolution/issues/29
    - https://github.com/capi-workgroup/decisions/issues/19

Return values
=============

The return value of a function must indicate whether an exception was set.

> It must not be necessary to use ``PyErr_Occurred`` to disambiguate.

Reviewer: Should we emphasise this more?

Author: I don't think so; the second redundant sentence should be fine.
Since it's a principle that guides the other guidelines, the reader will be fine if they skip it.

(This allows auto-generated API wrappers to avoid unnecessary special cases.)

As always, other exceptions can be added here with approval from the
C API working group.

Mind that infallibility is very often an implementation detail
that should not be exposed in the API. That is, some functions'
*current CPython implementations* cannot fail,
but we may want the function to fail (or warn) in the future.
Users should only skip error checking if the function's documentation
explicitly allows it.

.. _output argument:

Output arguments
================

.. (There's nothing to say for *arguments* in general, it's all under
   types or reference conting.
   If that changes, "Output arguments" should be a subsection of "Arguments")

Output arguments are pointers to memory that a function fills in.
Use these when a result from a function cannot be returned as the return value,
for example:

* ``-1`` or ``NULL`` is a valid (non-exceptional) result: for example,
  in ``PyDict_GetItemRef``.
* There are multiple results: for example, in ``PyUnicode_AsUTF8AndSize``.

Guidelines for output arguments:

* Functions must always fill in the output arguments. If an error

>   occurs or the rseult is not available, the output should typically be set

Reviewer 3: ```suggestion
  occurs or the result is not available, the output should typically be set
```

>   to NULL or zero.

Reviewer: ```suggestion
  to ``NULL`` or zero.
```

* When it might be useful for users to call a function but ignore an output,

>   allow passing NULL as the output argument.

Reviewer: ```suggestion
  allow passing ``NULL`` as the output argument.
```

* Ownership of a ``PyObject*`` result is transferred to the caller,
  as with return values.

>   [TODO: Link to guidelines about ownership & borrowing, when those are added]

Reviewer: ```suggestion

    .. TODO: Link to guidelines about ownership & borrowing, when those are added
```

Author: I need to do a formatting pass before turning this into a PEP. Before that, I prefer visible TODO reminders.

Since existing API does not necessarily follow these guidelines,
all of the above points should be explicitly mentioned in the
documentation of each function they're relevant to.

.. note::

    For background and discussions, see:

    - https://github.com/capi-workgroup/api-evolution/issues/32

.. _shadowing example:

Appendix A. Shadowing example
=============================

To provide a ``static inline`` equivalent to an exported function,
write something like:

Header:

.. code-block:: c

    static inline returntype
    _Py_Foo_impl(ARGS)
    {
        ...
    }

    PyAPI_FUNC(returntype) Py_Foo (ARGS);

>     #define Py_Foo _Py_Foo_impl

Reviewer 5: Some functions will want to expose the library function in the stable ABI for stable ABI reasons (at the expense of performance).  How is that normally done -- some sort of difference based on `Py_LIMITED_API`?  I realize that's a different concept, but it still might be good to have an example of a function exposed through the limited API, but exposed as static inline or macro through the non-limited API.

Author: Do this, but wrap the `static inline` function and the `#define` line in `#if !defined(Py_LIMITED_API)`.

I don't immediately know where to put this, so I'll [save your comment for later](https://github.com/capi-workgroup/api-evolution/issues/44).

Code:

.. code-block:: c

    // at the end (after all calls to Py_Foo):
    #undef Py_Foo

    returntype
    Py_Foo(ARGS)
    {
        return _Py_Foo_impl(ARGS);
    }

.. _return schemes:

Appendix B. Return value schemes
================================

Here are common schemes of how to encode return values.

*  Success or failure: return ``int``

   * ``0`` for success
   * ``-1``, with an exception set, for failure

*  Yes or no: return ``int``

>    *  ``0`` for ``false``
>    *  ``1`` for ``true``

Reviewer: Nit; match the description (yes before no) and visually separate the 1 and -1 case by arranging them 1, 0, -1:

```suggestion
   *  ``1`` for ``true``
   *  ``0`` for ``false``
```

   *  ``-1``, with an exception set, for failure

*  Lookup (“getattr”, “getitem” or “setdefault” style) functions: return
   ``int``; the lookup result is passed via an
   :ref:`output argument <output argument>`):

>    *  ``0`` for “not found”
>       (*result* is set to ``NULL`` or other zero/empty value)
>    *  ``1`` for “found” (*result* is set)

Reviewer: Ditto.

```suggestion
   *  ``1`` for “found” (*result* is set)
   *  ``0`` for “not found”
      (*result* is set to ``NULL`` or other zero/empty value)
```
